### PR TITLE
Add IngestPlugin api for plugins to add ingest processors

### PR DESCRIPTION
**Pull request by rjernst**
This change cuts ingest plugins over the new pull based plugins model.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/node/service/NodeService.java [hunk 1]
```diff
@@ -79,20 +79,13 @@ public NodeService(Settings settings, ThreadPool threadPool, MonitorService moni
...
+        this.scriptService = scriptService;
         clusterService.add(ingestService.getPipelineStore());
         clusterService.add(ingestService.getPipelineExecutionService());
     }
 
-    // can not use constructor injection or there will be a circular dependency
-    @Inject(optional = true)
-    public void setScriptService(ScriptService scriptService) {
-        this.scriptService = scriptService;
-        this.ingestService.buildProcessorsFactoryRegistry(scriptService, clusterService);
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/node/service/NodeService.java]: 👍

File: core/src/main/java/org/elasticsearch/plugins/IngestPlugin.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.ingest.Processor;
+import org.elasticsearch.ingest.TemplateService;
+import org.elasticsearch.script.ScriptService;
+
+/**
+ * An extension point for {@link Plugin} implementations to add custom ingest processors
+ */
+public interface IngestPlugin {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/plugins/IngestPlugin.java]: great!

File: core/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 3]
```diff
@@ -37,11 +39,20 @@
...
-        this.pipelineStore = new PipelineStore(settings);
+    public IngestService(Settings settings, ThreadPool threadPool,
+                         Environment env, ScriptService scriptService, List<IngestPlugin> ingestPlugins) {
+        final TemplateService templateService = new InternalTemplateService(scriptService);
+        Map<String, Processor.Factory> processorFactories = new HashMap<>();
+        for (IngestPlugin ingestPlugin : ingestPlugins) {
+            Map<String, Processor.Factory> newProcessors = ingestPlugin.getProcessors(env, scriptService, templateService);
+            for (Map.Entry<String, Processor.Factory> entry : newProcessors.entrySet()) {
+                if (processorFactories.put(entry.getKey(), entry.getValue()) != null) {
+                    throw new IllegalArgumentException("Ingest processor [" + entry.getKey() + "] is already registered");
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/IngestService.java]: maybe add a small test for this? (there was a small unit test for this: ProcessorsRegistryTests)

**Comment by martijnvg** [comment]
LGTM (left one minor note)

**Comment by rjernst** [comment]
Thanks @martijnvg, I added tests in f49ce8e

**Pull request merged**